packages/main/src/ComboBox.ts

```diff
@@ -0,0 +1,1240 @@
+		if (item.isGroupItem) {
+			let groupHasFilteredItems;
+
+			while (allItems[idx] && !allItems[idx].isGroupItem && !groupHasFilteredItems) {
+				groupHasFilteredItems = filteredItems.indexOf(allItems[idx]) !== -1;
+				idx++;
+			}
+
+			return groupHasFilteredItems;
+		}
+	}
+
+	_getFirstMatchingItem(current:string): ComboBoxItem | undefined {
+		const currentlyFocusedItem = this.items.find(item => item.focused === true);
+
+		if (currentlyFocusedItem?.isGroupItem) {
+			this.value = this.filterValue;
+			return;
+		}
+
+		const matchingItems: Array<ComboBoxItem> = (this._startsWithMatchingItems(current).filter(item => !item.isGroupItem) as Array<ComboBoxItem>);
+
+		if (matchingItems.length) {
+			return matchingItems[0];
+		}
+	}
+
+	_applyAtomicValueAndSelection(item: ComboBoxItem, filterValue: string, highlightValue: boolean) {
+		const value = (item && item.text) || "";
+
+		this.inner.value = value;
+		if (highlightValue) {
+			this.inner.setSelectionRange(filterValue.length, value.length);
+		}
+		this.value = value;
+	}
+
+	_selectMatchingItem() {
+		const currentlyFocusedItem = this.items.find(item => item.focused);
+		const shouldSelectionBeCleared = currentlyFocusedItem && currentlyFocusedItem.isGroupItem;
+
+		this._filteredItems = this._filteredItems.map(item => {
+			item.selected = !item.isGroupItem && (item.text === this.value) && !shouldSelectionBeCleared;
+			return item;
+		});
+	}
+
+	_fireChangeEvent() {
+		if (this.value !== this._lastValue) {
+			this.fireEvent("change");
+			this._lastValue = this.value;
+		}
+	}
+
+	_inputChange(e: Event) {
+		e.preventDefault();
+	}
+
+	_itemMousedown(e: MouseEvent) {
+		e.preventDefault();
+	}
+
+	_selectItem(e: CustomEvent<ClickEventDetail>) {
+		const listItem = e.detail.item as ComboBoxListItem;
+
+		this._selectedItemText = listItem.mappedItem.text;
+		this._selectionPerformed = true;
+
+		const sameItemSelected = this.value === this._selectedItemText;
+		const sameSelectionPerformed = this.value.toLowerCase() === this.filterValue.toLowerCase();
+
+		if (sameItemSelected && sameSelectionPerformed) {
+			this._fireChangeEvent(); // Click on an already typed, but not memoized value shouold also trigger the change event
+			return this._closeRespPopover();
+		}
+
+		this.value = this._selectedItemText;
+
+		if (!listItem.mappedItem.selected) {
+			this.fireEvent<ComboBoxSelectionChangeEventDetail>("selection-change", {
+				item: listItem.mappedItem,
+			});
+
+			this._selectionChanged = true;
+		}
+
+		this._filteredItems.map(item => {
+			item.selected = (item === listItem.mappedItem && !item.isGroupItem);
+			return item;
+		});
+
+		this._fireChangeEvent();
+		this._closeRespPopover();
+
+		// reset selection
+		this.inner.setSelectionRange(this.value.length, this.value.length);
+	}
+
+	_onItemFocus() {
+		this._itemFocused = true;
+	}
+
+	_announceSelectedItem(indexOfItem: number) {
+		const currentItem = this._filteredItems[indexOfItem] as ComboBoxItem;
```

Comment:
i would prefer to use https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates to determinate which combo box item is this one instead of casting all of them to comboboxitem